tests/std/tests/P2502R2_generator_promise/test.cpp

```diff
@@ -0,0 +1,242 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cassert>
+#include <coroutine>
+#include <cstddef>
+#include <cstdint>
+#include <cstdlib>
+#include <forward_list>
+#include <generator>
+#include <list>
+#include <memory>
+#include <memory_resource>
+#include <new>
+#include <ranges>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "range_algorithm_support.hpp"
+#include "test_generator_support.hpp"
+
+using namespace std;
+
+#pragma warning(disable : 28251) // Inconsistent annotation for 'new[]': this instance has no annotations.
+
+void* operator new[](size_t) {
+    abort();
+}
+
+void operator delete[](void*) noexcept {
+    abort();
+}
+
+template <class Promise, class... Args>
+concept HasOperatorNew = requires(Args&&... args) {
+    { Promise::operator new(forward<Args>(args)...) } -> same_as<void*>;
+};
+
+template <class Gen>
+struct generator_allocator {};
+
+template <class Ref, class V, class Alloc>
+struct generator_allocator<generator<Ref, V, Alloc>> {
+    using type = Alloc;
+};
+
+template <class Gen, class Range>
+    requires convertible_to<ranges::range_reference_t<Range&>, typename Gen::yielded>
+void test_yield_elements_of_range(typename Gen::promise_type& p) {
+    using Alloc = generator_allocator<Gen>::type;
+
+    {
+        using Awaitable = decltype(p.yield_value(ranges::elements_of{declval<Range&>()}));
+        static_assert(same_as<decltype(declval<Awaitable&>().await_ready()), bool>);
+    }
+
+    if constexpr (!is_void_v<Alloc>) {
+        using Awaitable = decltype(p.yield_value(ranges::elements_of{declval<Range&>(), declval<Alloc&>()}));
+        static_assert(same_as<decltype(declval<Awaitable&>().await_ready()), bool>);
+    }
+}
+
+template <class Gen, class Alloc2>
+void test_operator_new(typename Gen::promise_type& p, const Alloc2& alloc2 = {}) {
+    using Promise = Gen::promise_type;
+    using Alloc   = generator_allocator<Gen>::type;
+
+    // Test 'operator new(size_t)'
+    constexpr bool has_op_new1 = HasOperatorNew<Promise, size_t>;
+    static_assert(has_op_new1 == (same_as<Alloc, void> || default_initializable<Alloc>) );
+    if constexpr (has_op_new1) {
+        const size_t size = __STDCPP_DEFAULT_NEW_ALIGNMENT__;
+        void* const mem   = p.operator new(size);
+        assert(reinterpret_cast<intptr_t>(mem) % __STDCPP_DEFAULT_NEW_ALIGNMENT__ == 0);

```

Comment:
Suggestion: I would prefer to see `uintptr_t` here. (The less I have to think about negative values, the better.) Occurs below.